caikit/runtime/utils/servicer_util.py

```diff
@@ -260,6 +269,140 @@ def validate_data_model(
         validate_caikit_library_class_method_exists(caikit_Library_class, "to_proto")
 
 
+class ServicePackageStreamWrapper(DataStream):
+    """This class wraps up a DataStreamSourceBase derived class so that it can
+    be safely pickled to a subprocess using "spawn".
+
+    When using "spawn" to start a subprocess, a clean python process is created
+    and none of the runtime code leading to the creation is re-executed. For
+    most classes that are created at import time, this is fine, but for
+    dynamically created classes like those created with make_data_stream_source,
+    the class is not recreated on spawn and cannot be looked up.
+
+    To get around this, when unpickling this class, it falls back to performing
+    service generation dynamically if the class doesn't already exist
+    """
+
+    def __init__(self, stream: DataStreamSourceBase):
+        """Initialize with the data stream source to wrap
+
+        NOTE: We _intentionally_ don't initialize the base class here and
+            instead we forward
+        """
+        super().__init__(stream._generator)
+        self._stream = stream
+        update_wrapper(self, stream)
+
+    def __getstate__(self) -> Tuple[str, str, type, bytes]:
+        """Pickle as a tuple of primitive types
+
+        Returns:
+            stream_module_name (str): The name of the module that holds the
+                stream source class
+            stream_class_name (str): The name of the stream source class itself
+            stream_class_element_type (type): The type of the elements for this
+                stream source
+            content (bytes): The serialized bytes content of the underlying
+                stream source as a dataobject
+        """
+        return (
+            self._stream.__class__.__module__,
+            self._stream.__class__.__name__,
+            self._stream.__class__.ELEMENT_TYPE,
+            self._stream.__getstate__(),
+        )
+
+    def __setstate__(self, pickled: Tuple[str, str, type, bytes]):
+        """Unpickle the values from pickling and fall back to dynamic service
+        generation if needed.
+
+        Args:
+            pickled (Tuple[str, str, type, bytes]): The four elements created
+                when pickling this instance
+        """
+        stream_type_module, stream_type_name, element_type_class, stream_bytes = pickled
+        try:
+            mod = sys.modules[stream_type_module]
+            stream_type = getattr(mod, stream_type_name)
+        except (KeyError, AttributeError):
+            # Local
+            from caikit.runtime.service_factory import (  # pylint: disable=import-outside-toplevel
+                ServicePackageFactory,
+            )
+
+            ServicePackageFactory.get_service_package(
+                ServicePackageFactory.ServiceType.INFERENCE
+            )
+            ServicePackageFactory.get_service_package(
+                ServicePackageFactory.ServiceType.TRAINING
+            )
```

Comment:
I love it so much and yet hate it at the same time 🙃

Reply:
😬